core/src/avm2/value.rs

```diff
@@ -1305,4 +1331,708 @@ impl<'gc> Value<'gc> {
+                if (this.as_xml_object().is_some() || this.as_xml_list_object().is_some())
+                    && multiname.contains_public_namespace()
+                {
+                    return this.delete_property_local(activation, multiname);
+                }
+                Ok(false)
+            }
+        }
+    }
+
+    /// Same as delete_property, but constructs a public Multiname for you.
+    pub fn delete_public_property(
+        &self,
+        activation: &mut Activation<'_, 'gc>,
+        name: impl Into<AvmString<'gc>>,
+    ) -> Result<bool, Error<'gc>> {
+        let name = Multiname::new(activation.avm2().public_namespace, name);
+        self.delete_property(activation, &name)
+    }
+
+    /// Get this value's class, if it has one.
+    pub fn instance_of(&self, activation: &mut Activation<'_, 'gc>) -> Option<ClassObject<'gc>> {
+        match self {
+            Value::Object(o) => o.instance_of(),
+            Value::String(_) => Some(activation.avm2().classes().string),
+            Value::Number(_) => Some(activation.avm2().classes().number),
+            Value::Integer(_) => Some(activation.avm2().classes().int),
+            Value::Bool(_) => Some(activation.avm2().classes().boolean),
+            _ => None,
+        }
+    }
+
+    pub fn proto(&self, activation: &mut Activation<'_, 'gc>) -> Option<Object<'gc>> {
+        match self {
+            Value::Object(o) => o.proto(),
+            Value::String(_) => Some(activation.avm2().classes().string.prototype()),
+            Value::Number(_) | Value::Integer(_) => {
+                Some(activation.avm2().classes().number.prototype())
+            }
+            Value::Bool(_) => Some(activation.avm2().classes().boolean.prototype()),
+            _ => None,
+        }
+    }
+
+    /// Indicates whether or not a property exists on an object.
+    pub fn has_property(self, name: &Multiname<'gc>, activation: &mut Activation<'_, 'gc>) -> bool {
+        if self.has_own_property(name, activation) {
+            true
+        } else if let Some(proto) = self.proto(activation) {
+            proto.has_own_property(name)
+        } else {
+            false
+        }
+    }
+
+    pub fn is_sealed(&self) -> bool {
+        match self {
+            Value::Object(o) => o.base().is_sealed(),
+            // All primitives are sealed
+            Value::String(_) => true,
+            Value::Number(_) | Value::Integer(_) => true,
+            Value::Bool(_) => true,
+            _ => unreachable!("Called is_sealed on {self:?}"),
+        }
+    }
+
+    /// Call the object.
+    pub fn call(
+        self,
+        receiver: Value<'gc>,
+        arguments: &[Value<'gc>],
+        activation: &mut Activation<'_, 'gc>,
+    ) -> Result<Value<'gc>, Error<'gc>> {
+        if let Some(this) = self.as_object() {
+            this.call(receiver, arguments, activation)
+        } else {
+            Err("Primitive is not callable".into())
```

Comment:
Can this use the correct error?

Reply:
None of our tests hit it, so this can be done in a follow-up